File: service-monitor/src/main/scala/com/yahoo/vespa/service/monitor/SlobrokMonitor.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vespa.service.monitor

import com.yahoo.jrt.slobrok.api.Mirror.Entry
import com.yahoo.jrt.slobrok.api.{Mirror, SlobrokList}
import com.yahoo.jrt.{Transport, Supervisor}
import com.yahoo.vespa.service.monitor.SlobrokMonitor._
import scala.collection.convert.wrapAsJava._

/**
 * @author bakksjo
 */
class SlobrokMonitor {
  private val supervisor: Supervisor = new Supervisor(new Transport())
  private val slobrokList = new SlobrokList()
  private val mirror = new Mirror(supervisor, slobrokList)

  def setSlobrokConnectionSpecs(slobrokConnectionSpecs: Traversable[String]): Unit = {
    val slobrokConnectionSpecsJavaList: java.util.List[String] = slobrokConnectionSpecs.toList
    slobrokList.setup(slobrokConnectionSpecsJavaList.toArray(new Array[java.lang.String](0)))
  }

  def getRegisteredServices(): Map[SlobrokServiceName, SlobrokServiceSpec] = {
    if (!mirror.ready()) {
      return Map()
    }
    val mirrorEntries: Array[Entry] = mirror.lookup("**")
    mirrorEntries.map { mirrorEntry =>
      (SlobrokServiceName(mirrorEntry.getName), SlobrokServiceSpec(mirrorEntry.getSpec))
    }.toMap
  }

  def isRegistered(serviceName: SlobrokServiceName): Boolean = {
    mirror.lookup(serviceName.s).length != 0
  }

  def shutdown(): Unit = {
    mirror.shutdown()
  }
}

object SlobrokMonitor {
  case class SlobrokServiceName(s: String)
  case class SlobrokServiceSpec(s: String)
}

File: service-monitor/src/main/java/com/yahoo/vespa/service/monitor/SlobrokMonitor.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

package com.yahoo.vespa.service.monitor;

import com.yahoo.jrt.Supervisor;
import com.yahoo.jrt.Transport;
import com.yahoo.jrt.slobrok.api.Mirror;
import com.yahoo.jrt.slobrok.api.SlobrokList;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SlobrokMonitor {
    private final Supervisor supervisor = new Supervisor(new Transport());
    private final SlobrokList slobrokList = new SlobrokList();
    private final Mirror mirror = new Mirror(supervisor, slobrokList);

    void setSlobrokConnectionSpecs(List<String> slobroks) {
        slobrokList.setup(slobroks.toArray(new String[0]));
    }

    Map<SlobrokServiceName, SlobrokServiceSpec> getRegisteredServices() {
        if (!mirror.ready()) {
            return new HashMap<>();
        }

        Mirror.Entry[] mirrorEntries = mirror.lookup("**");
        return Arrays.asList(mirrorEntries).stream().collect(Collectors.toMap(
                entry -> new SlobrokServiceName(entry.getName()),
                entry -> new SlobrokServiceSpec(entry.getSpec())));
    }

    boolean isRegistered(SlobrokServiceName serviceName) {
        return mirror.lookup(serviceName.s()).length != 0;
    }

    void shutdown() {
        mirror.shutdown();
    }

    public static class SlobrokServiceName {
        private final String name;

        SlobrokServiceName(String name) {
            this.name = name;
        }

        // TODO: Fix spec
        public String s() {
            return name;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            SlobrokServiceName that = (SlobrokServiceName) o;

            return name.equals(that.name);
        }

        @Override
        public int hashCode() {
            return name.hashCode();
        }
    }

    public static class SlobrokServiceSpec {
        private final String spec;

        SlobrokServiceSpec(String spec) {
            this.spec = spec;
        }

        // TODO: Fix name
        public String s() {
            return spec;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            SlobrokServiceSpec that = (SlobrokServiceSpec) o;

            return spec.equals(that.spec);
        }

        @Override
        public int hashCode() {
            return spec.hashCode();
        }
    }
}
